ibmcloud-janitor/resources/serviceid.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"fmt"
+
+	identityv1 "github.com/IBM/platform-services-go-sdk/iamidentityv1"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/boskos/common/ibmcloud"
+	"sigs.k8s.io/boskos/ibmcloud-janitor/account"
+	"sigs.k8s.io/boskos/ibmcloud-janitor/utils"
+)
+
+type APIKey struct {
+	serviceIDName string
+	value         *string
+}
+
+func getAPIkeyName() (string, error) {
+	n, err := utils.RandomNumber()
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("test-key-%d", n), nil
+}
+
+// Lists and returns the target service ID
+func (s *ServiceID) fetchServiceID(account *string) (*identityv1.ServiceID, error) {
+	options := &identityv1.ListServiceIdsOptions{
+		AccountID: account,
+	}
+
+	list, _, err := s.ListServiceID(options)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to list the serviceIDs")
+	}
+
+	for _, sid := range list.Serviceids {
+		if *sid.Name == s.key.serviceIDName {
+			return &sid, nil
+		}
+	}
+	return nil, errors.Wrapf(err, "failed to fetch the target serviceID")
+}
+
+// Fetches the API keys under the target service ID and deletes them
+// followed by creating a new API key
+func (s *ServiceID) resetKeys(serviceID *identityv1.ServiceID) (*identityv1.APIKey, error) {
```

Comment:
@mkumatag, the failure scenarios can be handled as mentioned below. Please share your thoughts in case I missed anything here.
1. Accidental deletion of API key assigned to the resource when the state of the resource is

_Free/Busy_: 
- The triggered test would fails while authenticating with IBM Cloud since the API key no longer exists
-  The state of the resource is set to dirty once the test ends after which the janitor picks it up, cleans it and updates the resource with a new API Key followed by setting its state to free 
- Subsequent test runs will eventually pass by using the new API key for authentication

_Any other state_: 
- Reaper sets the state of the resource to dirty irrespective of the current state after a certain time period 
- Janitor cleans up the resource and updates the API key with a new value and sets it to state free after which it can be used by any test.


2. Janitor fails to create a new API key or updating the resource with the new key
- On failure, the resource would still be in a dirty state(can not be used by any project/owner) and janitor will be responsible for creating a new API key in subsequent runs before finally setting its state to free

Reply:
this sgtm

Reply:
@Amulyam24 In the second failure scenario, will the Janitor not be infinitely trying to Create a new API key? (Hypothetically, if the reason for APIKey creation failure is not resolved)
Hope Janitor or the Test has a timeout value.

Reply:
@Rajalakshmi-Girish, if there has been an unforeseen failure while trying to create a new API key, the resource would continuously be transitioning from cleaning <--> dirty state (set by reaper after 30m expiration time) thereby making it unusable by any test. I think resolving it would require probing the logs to find the actual cause and fixing it.

Reply:
@Amulyam24 Yes you are right, the second failure scenario happens with a resource in `dirty` state.

But if the cause for `create a new API key` failure is not in our scope of resolution, ( For example, if it is something to do with PowerVS) will Janitor infinitely keep trying to `create a new API Key`? or is there a timeout that Janitor has for dirty resources that it handles?

Reply:
> will Janitor infinitely keep trying to create a new API Key? 

Yes, until it is able to create one and set the resource to free state. 

> or is there a timeout that Janitor has for dirty resources that it handles?

There is no timeout set for Janitor if it continuously fails to clean up a dirty resource. It will keep looking for dirty resources until they are set to free.

Reply:
> Yes, until it is able to create one and set the resource to free state.

If Janitor is unable to create APIKey for quite some time, will it report the failure?

Should we consider the janitor to have a timeout for handling each dirty resource? Else, scenarios like the second one might lead to a situation where resource remains in `dirty` state and might not be reported.

Also, there might be a situation where all resources posted to `Boskos` might move to `dirty` state and eventually lead to test failure (as it would be unable to find a free resource to run the test).